sw/device/silicon_creator/lib/epmp_unittest.cc

```diff
@@ -0,0 +1,346 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/epmp.h"
+
+#include <vector>
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/csr.h"
+#include "sw/device/lib/base/testing/mock_csr.h"
+
+namespace epmp_unittest {
+using ::mock_csr::CsrTest;
+using ::testing::Test;
+
+/**
+ * PMP control registers.
+ */
+struct PmpCsrs {
+  std::vector<uint8_t> pmpcfg = std::vector<uint8_t>(kEpmpNumRegions, 0);
+  std::vector<uint32_t> pmpaddr = std::vector<uint32_t>(kEpmpNumRegions, 0);
+  uint64_t mseccfg = 0;
+
+  epmp_state_t State() const {
+    epmp_state_t state;
+    for (int i = 0; i < kEpmpNumRegions; ++i) {
+      bitfield_field32_t field = {.mask = 0xff,
+                                  .index = (static_cast<uint32_t>(i) % 4) * 8};
+      state.pmpcfg[i / 4] = bitfield_field32_write(
+          state.pmpcfg[i / 4], field, static_cast<uint32_t>(pmpcfg.at(i)));
```

Comment:
I might be missing something but why do you need to use `at()` (sizes and loop ranges use the same constant)?

Reply:
Just on the basis that I like bounds checking. I think it is better to use `at` more or less exclusively unless the code is performance sensitive.